src/walk-manifest.js

```diff
@@ -49,116 +50,197 @@ var parseManifest = function(content) {
+    parent.content = new Buffer(parent.content.toString().replace(uri, path.relative(path.dirname(parent.file), manifest.file)));
+  }
+
+  if (visitedUrls.includes(manifest.uri)) {
+    const manifestError = new Error('Trying to visit the same uri again; stuck in a cycle|' + manifest.uri);
+    return onError(manifestError, resources, callback);
+  }
+
+  request({url: manifest.uri, timeout: 1500}, function(error, response, body) {
+    if (error) {
+      const manifestError = new Error(error.message + '|' + manifest.uri);
+      return onError(manifestError, resources, callback);
+    }
+    if (response.statusCode !== 200) {
+      const manifestError = new Error(response.statusCode + '|' + manifest.uri);
+      return onError(manifestError, resources, callback);
+    }
+    // Only push manifest uris that get a non 200 and don't timeout
+    resources.push(manifest);
+    visitedUrls.push(manifest.uri);
+
+    manifest.content = body;
+
+    manifest.parsed = parseManifest(manifest.content);
+    manifest.parsed.segments = manifest.parsed.segments || [];
+    manifest.parsed.playlists = manifest.parsed.playlists || [];
+    manifest.parsed.mediaGroups = manifest.parsed.mediaGroups || {};
+
+    var playlists = manifest.parsed.playlists.concat(mediaGroupPlaylists(manifest.parsed.mediaGroups));
+    parseKey(basedir, decrypt, resources, manifest, parent, function(err, key) {
+      // SEGMENTS
+      manifest.parsed.segments.forEach(function(s, i) {
+        if (!s.uri) {
+          return;
+        }
+        // put segments in manifest-name/segment-name.ts
+        s.file = path.join(path.dirname(manifest.file), fsSanitize(path.basename(s.uri)));
+        if (!isAbsolute(s.uri)) {
+          s.uri = joinURI(path.dirname(manifest.uri), s.uri);
+        }
+        if (key) {
+          s.key = key;
+          s.key.iv = s.key.iv || new Uint32Array([0, 0, 0, manifest.parsed.mediaSequence, i]);
+        }
+        manifest.content = new Buffer(manifest.content.toString().replace(
+          s.uri,
+          path.relative(path.dirname(manifest.file), s.file)
+        ));
+        resources.push(s);
+      });
+
+      // SUB Playlists
+      // The reflect is used so we can still continue running even if one of the playlists is broken.
+      async.map(playlists, async.reflect(function(p, cb) {
+        if (!p.uri) {
+          return cb(null);
+        }
+        walkPlaylist({
+          decrypt,
+          basedir,
+          uri: p.uri,
+          parent: manifest,
+          manifestIndex: playlists.indexOf(p),
+          onError,
+          visitedUrls
+        }, cb);
+      }), function(err, reflectedResults) {
+        var results = [];
+        var errors = [];
+
+        reflectedResults.forEach(function(r) {
+          if (r.error) {
+            errors.push(r.error);
+          }
+          if (r.value) {
+            results.push(r.value)
+          }
+        });
+
+        var flattenedResource = [].concat.apply([], results);
```

Comment:
might be worth extracting this or creating a polyfill for `flat`

Reply:
if these arrays are being created above, why not just concat/flatten them there? Hopefully the reflected stuff can be washed away by more standard control flow stuff though.